### ISSUE TITLE
Pandas_input_fn slow, starving CPU/GPU

### ISSUE THREAD

**Issue by mellvinbaker** [tensorflow/tensorflow#13530]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: It is a customized version of the Deep & Wide example code. Fairly close to original code.
- **OS Platform and Distribution: Windows Server 2012 R2
- **TensorFlow installed from: nightly build WHL through pip (this was tried after numerous other versions, including install through pip)
- **TensorFlow version (use command below)**: b'unknown' 1.4.0-dev20170926
- **Python version**: 3.5 and 3.6
- **Bazel version (if compiling from source)**: Not compiling
- **CUDA/cuDNN version**: CUDA 8, CUDnn 6.1
- **GPU model and memory**: Tesla M60 GPU 8GB
- **Exact command to reproduce**:  See attached Script.
-For the record, the server vm has 8 xeon physical cores and 240 gb ram allocated. The CPU only machine is a new skylake i7 with 32gb ram.

### Describe the problem
To start, I submitted to stack overflow (https://stackoverflow.com/questions/46457476/tensorflow-pandas-input-fn-slow-starving-cpu-gpu) and have not been able to garner assistance after multiple edits to make sure it was framed correctly. I truly believe this is a bug since I am sticking so close to the example code, but if I have made a mistake I am deeply sorry to all of you.

I am working on a wide and deep model following the framework in the Tensorflow Wide and Deep tutorial (https://www.tensorflow.org/tutorials/wide_and_deep). Model works fine when built the old way (load entire dataset from pandas, convert to tensors, feed in input_fn) which is ok for running on a CPU. 

However, to make it work on the GPU the dataset is too large to fit into GPU memory, so batching is necessary. I tried using the pandas_input_fn to batch data to the video card and noticed I get spikes of activity followed by long lulls while the next batch is prepared. The odd thing is, this happens even if I run it on a machine with CPU only. The lulls are almost the exact same length, so it isn't simply the video card crushing through a simple model faster than the proc can deliver it. It seems like it is always waiting to begin loading the next batch until the last one is done training. 

(If this function simply cannot be used in this way, can we get an example of Deep and Wide using the dataset API? or a manual build of deep and wide using layers and queues? At the moment, the example code for the dataset api using make_one_shot_iterator for canned estimators doesn't run.)

I increased the complexity of the model to make sure it wasn't too easy to compute and still have the same issue. I have tried increasing the number of threads allocated to pandas_input_fn, I have tried increasing the queue size to far larger than seems reasonable (10x dataset size) which helps a bit, but not much. I am not sure if the slowdown is when it is queueing or de-queueing, but I have been unable to solve the issue after two weeks of troubleshooting. The data I am working with is 117 columns, 400k rows.

I have created a generic script that generates fake values to simulate the problem. However, there are far fewer fake columns than real ones, so the gap between steps is not nearly as long, but still noticeable. Code attached.

--

### Source code / logs
attached
[pandas_input_example.txt](https://github.com/tensorflow/tensorflow/files/1363335/pandas_input_example.txt)

**Comment by aselle** [comment]
@mrry, could you make some suggestions?

**Comment by apbodnar** [comment]
I too have been dealing with extremely low (<10%) GPU usage when using `pandas_input_fn` with a `DNNRegressor` and Titan Xp

**Comment by mellvinbaker** [comment]
I have not addressed usage itself yet because I assumed it was due to the balance of loading data. I couldn't dial in that balance with the GPU taking such long pauses. That said, my GPU usage was also extremely low, even when I seemed to be filling memory up (growth set to true, etc.) I never got above 12%. Still, I am much more worried about the batching speed issue. This bug, if it is indeed a bug, means we are still stuck loading tabular data via tfrecord/protobuf if we want decent performance. Its a problem because its more than 90% of our ML use cases (image data is fun, but a huge portion of ML is just strait up data). That said, I'm not complaining overall, Tensorflow is an amazing achievement and it gets better every day. Thanks to all the devs volunteering on this project.

**Comment by mrry** [comment]
@mellvinbaker Looking at your example code, I noticed that the `batch_size` is 200000 (i.e. the full size of the `DataFrame`). Is that what you'd typically use in your application?

Given that the `pandas_input_fn()` is based on a `FeedingQueueRunner`, it would have to feed 200000 elements to build such a batch. At ~20us per feed, I'm not surprised that it stalls for a noticeable amount of time.

**Comment by mellvinbaker** [comment]
The dataframe is 400,000 rows, not 200,000 and those 200,000 rows make for about ~3gb of data. I am not sure how that is any different than a 3GB array of image values. I would not think that batching 3gb of numbers to a GPU would be outside the scope of pandas_input_fn, but maybe I misunderstand its purpose (or its use, both are very possible). When would you ever need to batch a dataset of say 500mb to a tesla M60? The whole thing would fit in memory, thus avoiding the need for batching at all.

The typical use would be to pull about 2,000,000 rows from SQL, batch them into the most efficient size for the GPU, and run the model. At the moment, I can run the entire 2,000,000 through an estimator if I use CPU only because I have ~128gb of ram on the server. That runs in half the time of batching that same 2,000,000 rows to the GPU, regardless of batch size. 

I have tried batch sizes ranging from 20k to 200k and all have the same problem. When one batch is done, the GPU sits idle waiting for more.

**Comment by mellvinbaker** [comment]
FYI, I have rebuilt my input functions to see if TFRecord would work better. While it does do a better job of keeping the processor fed without lag (still a little between steps, but I haven't even optimized, still doing 1 file 1 thread etc.) it isn't a solution for this type of data. It makes sense for pictures, which change very little. But if you are working with a dataset that needs feature engineering it can be a bit iterative. 
With 400k rows taking ~8 hours to serialize to TFRecord, we entirely lose any advantage gained by using GPU.

Thanks again guys for looking into issues like this and for all the hard work.

**Comment by Goorman** [comment]
Checking in to say that i encountered the same kibd of problem. 2kk objects with 300 features, feeded to wideanddeep estimator through TFRecordDataset, cpu is overloaded at 800%(8 cores) while titan gpu isat 10% utilisation. Tried on various batch sizes from 64 to 20k.

Is there sonethong i am doing wrong or is there any updates about this issue?

**Comment by mrry** [comment]
Reassigning this to @rohan100jain, who has been working on a GPU prefetching optimization for `tf.data` that should fix the utilization problem.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by 5ean** [comment]
had the same issue, really want to know how to improve the performance.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rohan100jain** [comment]
Just wanted to ping here and mention that I'm indeed working on a GPU prefetching optimization that should help parallelize loading of batches much more and hence improve performance. The work is nearly done - I'm just testing and making sure it functions correctly before publishing some instructions as to how to use it. Please stay tuned for a week or so as I solidify this more. Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 79 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 109 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gongzg** [comment]
@rohan100jain any update for this performance issue? From the thread above, I assume there is already a solution, wonder when will it be published? Thanks.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tkainazarov** [comment]
It is not only pandas, direct reading from csv (TextLineDataset), DNNRegressor still underutilising GPU.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rohan100jain** [comment]
Please try out tf.data.experimental.copy_to_device("/gpu:0").prefetch(1) to get prefetching onto the GPU.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by Quisaxaderak** [comment]
Sorry, I just saw the response. Where would I put this, is it a runconfig parameter, part of pandas_input_fn, a decorator at the top of the script? Or does this only apply to text_line_dataset?

**Comment by harshini-gadige** [comment]
> Sorry, I just saw the response. Where would I put this, is it a runconfig parameter, part of pandas_input_fn, a decorator at the top of the script? Or does this only apply to text_line_dataset?

@rohan100jain  Could you please answer this ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mathlf2015** [comment]
I have the same problem, the utilization of gpu is too low，really want to know how to improve the performance.

**Comment by mehdi-beng** [comment]
Having same issue with CSV input function using TextLineDataset (w/ decodeCsv call). Read speeds are extremely low. I am seeing 2MB/s max even on SSD drive

**Comment by rohan100jain** [comment]
The prefetching currently works only for tf.data, so you'll have to replace your pandas_input_fn with a dataset. I believe you can use SQLDAtaset (https://www.tensorflow.org/versions/r1.12/api_docs/python/tf/data/experimental/SqlDataset?hl=en) to construct one of these.

Once you have a dataset (lets call it sql_ds) with the information, you can do the following

def input_fn():
  sql_ds = <create_sql_dataset>
  ds = sql_ds.apply(tf.data.experimental.copy_to_device("/gpu:0")).prefetch(1)
  return ds

Estimators allow returning datatsets in their input fn and things should flow from there.

For the CSV input fn and other cases, just use different input dataset (CSVDataset for example: https://www.tensorflow.org/api_docs/python/tf/data/experimental/CsvDataset?hl=en)

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by Jumonji** [comment]
Same problem with pandas_input_fn. Less than 10% GPU utilization on my brand new RTX 2080 TI... I spent all that money and I can process just as fast on a multi-core AMD CPU??? I don't get it. How are the "big boys" using these cards to get speedup?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by Jumonji** [comment]
> It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

Well, I've been focusing on learning 2.0 so I haven't focused on this - but still not getting over 15% GPU utilization (with very short bursts to 25%) no matter what I do.  Is there any example I can download to demonstrate full GPU utilization? That would be really helpful.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!
